doc/proposals/sdk-integration-with-olm.md

```diff
@@ -75,14 +88,126 @@ Initial PR: https://github.com/operator-framework/operator-sdk/pull/1912
+The idea is to create a `Deployment` containing the latest `operator-registry` [image][registry-image] to initialize a bundle database and run a registry server serving that database using binaries contained in the image. The `Deployment` will contain volume mounts from a `ConfigMap` containing bundle files and a package manifest for an Operator. Using manifest data in the `ConfigMap` volume source, the registry initializer can build a local database and serve that database through the `Service`. OLM-specific resources created by the SDK or supplied by a user, described below, will establish communication between this registry server and OLM.
 
 #### OLM resources
 
-OLM understands `operator-registry` servers and served data through several objects. A [`CatalogSource`][olm-catalogsource] specifies how to communicate with a registry server. A [`Subscription`][olm-subscription] links a particular CSV channel to a `CatalogSource`, indicating from which `CatalogSource` OLM should pull an Operator. Another OLM resource that _may_ be required is an [`OperatorGroup`][olm-operatorgroup], which provides Operator namespacing information to OLM; OLM creates two `OperatorGroup`'s by default, one of which can be used for globally scoped Operators.
+OLM understands `operator-registry` servers and served data through several objects. A [`CatalogSource`][olm-catalogsource] specifies how to communicate with a registry server. A [`Subscription`][olm-subscription] links a particular CSV channel to a `CatalogSource`, indicating from which `CatalogSource` OLM should pull an Operator. Another OLM resource that _may_ be required is an [`OperatorGroup`][olm-operatorgroup], which provides Operator namespacing information to OLM. OLM creates a globally-scoped `OperatorGroup` by default, which can be used for globally-scoped Operators.
 
 These resources can be created from bundle data with minimal user input. They can also be created from manifests defined by the user; however, the SDK cannot make guarantees that user-defined manifests will work as expected.
 
+#### OperatorGroups and tenancy requirements
+
+[`OperatorGroup`][olm-operatorgroup]'s configure CSV tenancy in multiple
+namespaces in a cluster. Each Operator must be a
+[member][olm-operatorgroup-membership] with one `OperatorGroup` resource in
+the cluster, which defines a set of namespaces the CSV can exist in. A CSV's
+`installModes` determine what [type][olm-operatorgroup-installmodes] of
+`OperatorGroup` it can be a member of.
+
+No two `OperatorGroup`'s can exist in the same namespace, and a CSV with
+membership in an `OperatorGroup` of a type it does not support (determined
+by `installModes`) will transition to a failure state.
+
+Given these rules and constraints, Operator developers may have a tough time
```

Comment:
The document from here on is not wrong, but I think it should probably do more to indicate that installmodes should describe the way the operator works and are not things that can be changed "after the fact" without rewriting/etc.

If the operator starts up and watches all namespaces, it should be AllNamespace, and should go into an OperatorGroup that watches all namespaces.

If the operator starts up and watches its own namespace, it should be OwnNamespace and go into an OperatorGroup that watches its own namespace only.

If the operator starts up and watches a single namespace based on an env var, and that env var is wired up to project the `olm.targetNamespaces` annotation from the deployment, then it should go into a SingleNamespace OperatorGroup.

Likewise, multinamespace mode and the configuration thereof is fundamental to the way the operator starts up and generating an operatorgroup with multiple namespaces will do nothing if the operator doesn't support watching n namespaces.

Operators can also support one or all of these, depending on how it is written.

A lot of this can be determined based on the properties of the operator itself:

1. If it uses the downward API to read the namespace annotations, then it can potentially support Single and Multi namespace modes (the inverse might make more sense - if a deployment for an operator does not read from the annotations, then it cannot possible satisfy single or multi namespace installmode requirements).
2. If the operator can only start up control loops per a configured namespace value (be that a specific namespace or AllNamespaces `""`), then it cannot support MultiNamespace

The current proposal is fine and makes sense! But I'm wondering if there's a way that the sdk can "just know" what operatorgroup to make and what installmodes to make more directly, because of knowledge of what namespaces can be watched by the operator.

Reply:
What about establishing a convention that is based on the order of priority of the following conditions:

If AllNamespace is supported, create an OperatorGroup that watches all namespaces.
If SingleNamespace is supported, create an OperatorGroup that watches the namespace the Operator is deployed in.
If OwnNamespace is supported, same as above.
If MultiNamespace is supported, same as above.